

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to cobilib’s documentation! &mdash; cobilib 1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="cobilib 1.0 documentation" href="#" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="#">cobilib 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="welcome-to-cobilib-s-documentation">
<h1>Welcome to cobilib&#8217;s documentation!<a class="headerlink" href="#welcome-to-cobilib-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul class="simple">
</ul>
</div>
<span class="target" id="module-cobilib"></span><div class="section" id="brief">
<h2>Brief<a class="headerlink" href="#brief" title="Permalink to this headline">¶</a></h2>
<p>A library for analysing codon usage bias with the quasispecies model.</p>
</div>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="routines">
<h2>Routines<a class="headerlink" href="#routines" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt>All included routines::</dt>
<dd>CodonUsage
Fitnessfunction
Model
__builtins__
__doc__
__file__
__name__
__package__
c_codon_mut_dist
calculate_CAI_dic
calculate_NC
calculate_RF
calculate_RF_dic
calculate_RSCU
calculate_RSCU_dic
change_amino_acid_code
codon_hist_index
codon_index
codon_mut_dist
codon_table
codons
compute_distance
compute_optimum
compute_optimum_two_step
compute_steady_state
config_from_file
decomposition
euclidean_distance
hessian_em
highly_expressed_genes_by_file
highly_expressed_genes_by_id
init_fitnessfunctions
init_fitnessmatrix_for_amino_acids
init_fitnessmatrix_for_codons
init_list_of_genes
is_ambig_codon
kl_distance
load_fasta
load_fasta_from_url
load_genbank
load_plaintext
make_codon_histogram
make_codon_histogram_dic
make_codon_histogram_dic_combined
make_evolmatrix
make_jc69_mutationmatrix
make_mutationmatrix
mds_from_fasta
minimize
number_of_codons
optimize
optimize_dic
optimize_sequence
parametric_run_from_config
plot_all_reduction_methods
plot_mds
plot_pca
py_codon_mut_dist
relative_cosine_distance
relative_euclidean_distance
relative_hellinger_distance
relative_jeffrey_divergence
relative_minkowski_distance
remove_stopcodons_and_one_codon_amino_acids
remove_stopcodons_and_one_codon_amino_acids_dic
run_model
sample_run
setup_parser
transition</dd>
</dl>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="cobilib.Fitnessfunction">
<em class="property">class </em><tt class="descclassname">cobilib.</tt><tt class="descname">Fitnessfunction</tt><big>(</big><em>description</em>, <em>parameter</em>, <em>values</em>, <em>interpolation</em>, <em>filename=''</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#Fitnessfunction"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.Fitnessfunction" title="Permalink to this definition">¶</a></dt>
<dd><p>A Class for fitness functions!</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.c_codon_mut_dist">
<tt class="descclassname">cobilib.</tt><tt class="descname">c_codon_mut_dist</tt><big>(</big><em>a</em>, <em>b</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#c_codon_mut_dist"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.c_codon_mut_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the number of transitions, transversions and staying-const when one codon mutates to another.
and is implemented in C with scipy.weave. A pure Python version for easier readability is implemented via
py_codon_mut_dist.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>a: string with codon 1</strong> :</p>
<p><strong>b: string with codon 2</strong> :</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>List with three doubles `results` with `results[0]` containing the number of unchanged nucleotides,</strong> :</p>
<p class="last"><strong>`results[1]` number of transitions and `results[2]` number of transversions.</strong> :</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="cobilib.calculate_RF">
<tt class="descclassname">cobilib.</tt><tt class="descname">calculate_RF</tt><big>(</big><em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#calculate_RF"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.calculate_RF" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates relative codon frequency for a codon</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.calculate_RF_dic">
<tt class="descclassname">cobilib.</tt><tt class="descname">calculate_RF_dic</tt><big>(</big><em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#calculate_RF_dic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.calculate_RF_dic" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates relative codon frequency for each gene in codon histogram</p>
<p>Test against run of testdata2.ffn and handcalculated for amino acid A (codons 52 - 55) should give 0.24,0.3066,0.12,0.33</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">calculate_RF_dic</span><span class="p">(</span>  <span class="p">(</span><span class="n">make_codon_histogram_dic</span><span class="p">(</span> <span class="n">load_fasta</span><span class="p">(</span><span class="s">&quot;testdata2.ffn&quot;</span><span class="p">)))[</span><span class="mi">0</span><span class="p">]</span> <span class="p">)</span> <span class="c"># doctest : +NORMALIZE_WHITESPACE</span>
<span class="go">{&#39;fid|18348942|locus|VBIEscCol44059_0001|&#39;: array([ 0.375     ,  0.625     ,  0.14102564,  0.11538462,  0.15555556,</span>
<span class="go">        0.26666667,  0.13333333,  0.2       ,  0.57894737,  0.42105263,</span>
<span class="go">        0.        ,  0.        ,  0.25      ,  0.75      ,  1.        ,</span>
<span class="go">        1.        ,  0.07692308,  0.14102564,  0.03846154,  0.48717949,</span>
<span class="go">        0.12      ,  0.16      ,  0.04      ,  0.68      ,  0.58333333,</span>
<span class="go">        0.41666667,  0.36363636,  0.63636364,  0.3902439 ,  0.43902439,</span>
<span class="go">        0.04878049,  0.09756098,  0.68421053,  0.31578947,  0.        ,</span>
<span class="go">        1.        ,  0.14285714,  0.60714286,  0.        ,  0.25      ,</span>
<span class="go">        0.58823529,  0.41176471,  0.65517241,  0.34482759,  0.11111111,</span>
<span class="go">        0.13333333,  0.        ,  0.02439024,  0.29508197,  0.2295082 ,</span>
<span class="go">        0.09836066,  0.37704918,  0.24      ,  0.30666667,  0.12      ,</span>
<span class="go">        0.33333333,  0.675     ,  0.325     ,  0.72340426,  0.27659574,</span>
<span class="go">        0.35714286,  0.35714286,  0.16071429,  0.125     ]), &#39;fid|129049020348348942|locus|VBIEscCol44059_0001|&#39;: array([ 0.375     ,  0.625     ,  0.14102564,  0.11538462,  0.15555556,</span>
<span class="go">        0.26666667,  0.13333333,  0.2       ,  0.57894737,  0.42105263,</span>
<span class="go">        0.        ,  0.        ,  0.25      ,  0.75      ,  1.        ,</span>
<span class="go">        1.        ,  0.07692308,  0.14102564,  0.03846154,  0.48717949,</span>
<span class="go">        0.12      ,  0.16      ,  0.04      ,  0.68      ,  0.58333333,</span>
<span class="go">        0.41666667,  0.36363636,  0.63636364,  0.3902439 ,  0.43902439,</span>
<span class="go">        0.04878049,  0.09756098,  0.68421053,  0.31578947,  0.        ,</span>
<span class="go">        1.        ,  0.14285714,  0.60714286,  0.        ,  0.25      ,</span>
<span class="go">        0.58823529,  0.41176471,  0.65517241,  0.34482759,  0.11111111,</span>
<span class="go">        0.13333333,  0.        ,  0.02439024,  0.29508197,  0.2295082 ,</span>
<span class="go">        0.09836066,  0.37704918,  0.24      ,  0.30666667,  0.12      ,</span>
<span class="go">        0.33333333,  0.675     ,  0.325     ,  0.72340426,  0.27659574,</span>
<span class="go">        0.35714286,  0.35714286,  0.16071429,  0.125     ])}</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="cobilib.calculate_RSCU">
<tt class="descclassname">cobilib.</tt><tt class="descname">calculate_RSCU</tt><big>(</big><em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#calculate_RSCU"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.calculate_RSCU" title="Permalink to this definition">¶</a></dt>
<dd><p>returns codon_rscu for a gene</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.calculate_RSCU_dic">
<tt class="descclassname">cobilib.</tt><tt class="descname">calculate_RSCU_dic</tt><big>(</big><em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#calculate_RSCU_dic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.calculate_RSCU_dic" title="Permalink to this definition">¶</a></dt>
<dd><p>returns codon_rscu for each gene in codon histogram</p>
<p>test like calculate_rf and checked against genomes.urv.es/optimizer</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">calculate_RSCU_dic</span><span class="p">(</span>  <span class="p">(</span><span class="n">make_codon_histogram_dic</span><span class="p">(</span> <span class="n">load_fasta</span><span class="p">(</span><span class="s">&quot;testdata2.ffn&quot;</span><span class="p">)))[</span><span class="mi">0</span><span class="p">]</span> <span class="p">)</span>
<span class="go">{&#39;fid|18348942|locus|VBIEscCol44059_0001|&#39;: array([ 0.75      ,  1.25      ,  0.84615385,  0.69230769,  0.93333333,</span>
<span class="go">        1.6       ,  0.8       ,  1.2       ,  1.15789474,  0.84210526,</span>
<span class="go">        0.        ,  0.        ,  0.5       ,  1.5       ,  3.        ,</span>
<span class="go">        1.        ,  0.46153846,  0.84615385,  0.23076923,  2.92307692,</span>
<span class="go">        0.48      ,  0.64      ,  0.16      ,  2.72      ,  1.16666667,</span>
<span class="go">        0.83333333,  0.72727273,  1.27272727,  2.34146341,  2.63414634,</span>
<span class="go">        0.29268293,  0.58536585,  2.05263158,  0.94736842,  0.        ,</span>
<span class="go">        1.        ,  0.57142857,  2.42857143,  0.        ,  1.        ,</span>
<span class="go">        1.17647059,  0.82352941,  1.31034483,  0.68965517,  0.66666667,</span>
<span class="go">        0.8       ,  0.        ,  0.14634146,  1.18032787,  0.91803279,</span>
<span class="go">        0.39344262,  1.50819672,  0.96      ,  1.22666667,  0.48      ,</span>
<span class="go">        1.33333333,  1.35      ,  0.65      ,  1.44680851,  0.55319149,</span>
<span class="go">        1.42857143,  1.42857143,  0.64285714,  0.5       ]), &#39;fid|129049020348348942|locus|VBIEscCol44059_0001|&#39;: array([ 0.75      ,  1.25      ,  0.84615385,  0.69230769,  0.93333333,</span>
<span class="go">        1.6       ,  0.8       ,  1.2       ,  1.15789474,  0.84210526,</span>
<span class="go">        0.        ,  0.        ,  0.5       ,  1.5       ,  3.        ,</span>
<span class="go">        1.        ,  0.46153846,  0.84615385,  0.23076923,  2.92307692,</span>
<span class="go">        0.48      ,  0.64      ,  0.16      ,  2.72      ,  1.16666667,</span>
<span class="go">        0.83333333,  0.72727273,  1.27272727,  2.34146341,  2.63414634,</span>
<span class="go">        0.29268293,  0.58536585,  2.05263158,  0.94736842,  0.        ,</span>
<span class="go">        1.        ,  0.57142857,  2.42857143,  0.        ,  1.        ,</span>
<span class="go">        1.17647059,  0.82352941,  1.31034483,  0.68965517,  0.66666667,</span>
<span class="go">        0.8       ,  0.        ,  0.14634146,  1.18032787,  0.91803279,</span>
<span class="go">        0.39344262,  1.50819672,  0.96      ,  1.22666667,  0.48      ,</span>
<span class="go">        1.33333333,  1.35      ,  0.65      ,  1.44680851,  0.55319149,</span>
<span class="go">        1.42857143,  1.42857143,  0.64285714,  0.5       ])}</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="cobilib.change_amino_acid_code">
<tt class="descclassname">cobilib.</tt><tt class="descname">change_amino_acid_code</tt><big>(</big><em>code_name</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#change_amino_acid_code"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.change_amino_acid_code" title="Permalink to this definition">¶</a></dt>
<dd><p>Various organisms have different genetic codes. Given the <cite>code_name</cite> which must be a member of <cite>genetic_codes</cite>
the global dictionaries for translating codons into amino acids are reset with the genetic code you want.:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Standard Code&#39;</span><span class="p">]</span>                                        <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Vertebrate Mitochondrial Code&#39;</span><span class="p">]</span>                        <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNKKSS**VVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Yeast Mitochondrial Code&#39;</span><span class="p">]</span>                             <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWTTTTPPPPHHQQRRRRIIMMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Mold, Protozoan, and Coelenterate Mitochondrial Code&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Invertebrate Mitochondrial Code&#39;</span><span class="p">]</span>                      <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNKKSSSSVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Ciliate, Dasycladacean and Hexamita Nuclear Code&#39;</span><span class="p">]</span>     <span class="o">=</span> <span class="s">&#39;FFLLSSSSYYQQCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Echinoderm and Flatworm Mitochondrial Code&#39;</span><span class="p">]</span>           <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNNKSSSSVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Euplotid Nuclear Code&#39;</span><span class="p">]</span>                                <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCCWLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Bacterial, Archaeal and Plant Plastid Code&#39;</span><span class="p">]</span>           <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Alternative Yeast Nuclear Code&#39;</span><span class="p">]</span>                       <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CC*WLLLSPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Ascidian Mitochondrial Code&#39;</span><span class="p">]</span>                          <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNKKSSGGVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Alternative Flatworm Mitochondrial Code&#39;</span><span class="p">]</span>              <span class="o">=</span> <span class="s">&#39;FFLLSSSSYYY*CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNNKSSSSVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;The Blepharisma Nuclear Code&#39;</span><span class="p">]</span>                             <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY*QCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;Chlorophycean Mitochondrial Code&#39;</span><span class="p">]</span>                         <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY*LCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;Trematode Mitochondrial Code&#39;</span><span class="p">]</span>                             <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNNKSSSSVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;Scenedesmus obliquus mitochondrial Code&#39;</span><span class="p">]</span>                  <span class="o">=</span> <span class="s">&#39;FFLLSS*SYY*LCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;Thraustochytrium Mitochondrial Code&#39;</span><span class="p">]</span>                      <span class="o">=</span> <span class="s">&#39;FF*LSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#39;</span>
<span class="n">genetic_codes</span><span class="p">[</span><span class="s">&#39;Pterobranchia mitochondrial code&#39;</span><span class="p">]</span>                         <span class="o">=</span> <span class="s">&#39;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSSKVVVVAAAADDEEGGGG&#39;</span>
</pre></div>
</div>
<dl class="docutils">
<dt>code_name:str</dt>
<dd>the name of the code which must be a key of <cite>genetic_codes</cite></dd>
</dl>
<p class="rubric">Examples</p>
<p>After changing the genetic code to something different than the standard code, the global amino_acids variable should
have changed to the new code.
&gt;&gt;&gt; _=change_amino_acid_code(&#8216;The Mold, Protozoan, and Coelenterate Mitochondrial Code&#8217;)
&gt;&gt;&gt; amino_acids
&#8216;FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG&#8217;</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.codon_mut_dist">
<tt class="descclassname">cobilib.</tt><tt class="descname">codon_mut_dist</tt><big>(</big><em>a</em>, <em>b</em><big>)</big><a class="headerlink" href="#cobilib.codon_mut_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the number of transitions, transversions and staying-const when one codon mutates to another.
and is implemented in C with scipy.weave. A pure Python version for easier readability is implemented via
py_codon_mut_dist.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>a: string with codon 1</strong> :</p>
<p><strong>b: string with codon 2</strong> :</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>List with three doubles `results` with `results[0]` containing the number of unchanged nucleotides,</strong> :</p>
<p class="last"><strong>`results[1]` number of transitions and `results[2]` number of transversions.</strong> :</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="cobilib.compute_steady_state">
<tt class="descclassname">cobilib.</tt><tt class="descname">compute_steady_state</tt><big>(</big><em>evolmatrix</em>, <em>aa_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#compute_steady_state"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.compute_steady_state" title="Permalink to this definition">¶</a></dt>
<dd><p>The steady state is the eigenvector belonging to the largest eigenvalue of the evolution matrix for a specific amino acid</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.hessian_em">
<tt class="descclassname">cobilib.</tt><tt class="descname">hessian_em</tt><big>(</big><em>hist</em>, <em>n_neighbors=10</em>, <em>n_components=2</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#hessian_em"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.hessian_em" title="Permalink to this definition">¶</a></dt>
<dd><p>docstring for hessian_em</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.highly_expressed_genes_by_file">
<tt class="descclassname">cobilib.</tt><tt class="descname">highly_expressed_genes_by_file</tt><big>(</big><em>filename</em>, <em>hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#highly_expressed_genes_by_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.highly_expressed_genes_by_file" title="Permalink to this definition">¶</a></dt>
<dd><p>loads a list of highly expressed genes and returns and returns an index where 0 if no heg and 1 if heg is returned.
the format used is that from ecai/heg and it tries to find the id in the description of the histogram keys</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.highly_expressed_genes_by_id">
<tt class="descclassname">cobilib.</tt><tt class="descname">highly_expressed_genes_by_id</tt><big>(</big><em>id_list</em>, <em>hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#highly_expressed_genes_by_id"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.highly_expressed_genes_by_id" title="Permalink to this definition">¶</a></dt>
<dd><p>not implemented right yet</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.init_fitnessfunctions">
<tt class="descclassname">cobilib.</tt><tt class="descname">init_fitnessfunctions</tt><big>(</big><em>config=None</em>, <em>filenames=None</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#init_fitnessfunctions"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.init_fitnessfunctions" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads fitnessfunctions from file. Either config from file oder list of filenames.</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.init_fitnessmatrix_for_amino_acids">
<tt class="descclassname">cobilib.</tt><tt class="descname">init_fitnessmatrix_for_amino_acids</tt><big>(</big><em>config=None, filenames=None, amino_acid_order=['A', 'R', 'N', 'D', 'C', 'E', 'Q', 'G', 'H', 'I', 'K', 'L', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V']</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#init_fitnessmatrix_for_amino_acids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.init_fitnessmatrix_for_amino_acids" title="Permalink to this definition">¶</a></dt>
<dd><p>Load fitnessmatrix form file. Either config from file or list of filenames</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.is_ambig_codon">
<tt class="descclassname">cobilib.</tt><tt class="descname">is_ambig_codon</tt><big>(</big><em>codon</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#is_ambig_codon"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.is_ambig_codon" title="Permalink to this definition">¶</a></dt>
<dd><p>Tests whether the string <cite>codon</cite> contains an ambigous letter that is found in <cite>ambig_fasta_chars</cite></p>
<dl class="docutils">
<dt>codon:str</dt>
<dd>nucleotide sequence</dd>
</dl>
<p class="rubric">Examples</p>
<p>The table of codons, <cite>codons</cite> should not contain any ambigous codons
&gt;&gt;&gt; map(is_ambig_codon,codons)
[False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False]</p>
<p>However, the string &#8216;aay&#8217; should
&gt;&gt;&gt; is_ambig_codon(&#8216;aay&#8217;)
True</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.isomap">
<tt class="descclassname">cobilib.</tt><tt class="descname">isomap</tt><big>(</big><em>hist</em>, <em>n_neighbors=10</em>, <em>n_components=2</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#isomap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.isomap" title="Permalink to this definition">¶</a></dt>
<dd><p>docstring for isomap</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.lle">
<tt class="descclassname">cobilib.</tt><tt class="descname">lle</tt><big>(</big><em>hist</em>, <em>n_neighbors=10</em>, <em>n_components=2</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#lle"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.lle" title="Permalink to this definition">¶</a></dt>
<dd><p>docstring for lle</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.load_fasta">
<tt class="descclassname">cobilib.</tt><tt class="descname">load_fasta</tt><big>(</big><em>filename</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#load_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.load_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads a fasta file with filename as arg. Returns generator object with list of genes</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>filename</strong> : str</p>
<blockquote>
<div><p>Path to the file to load</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>fasta</strong> : Seq</p>
<blockquote class="last">
<div><p>Parser from Biopython</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>If everything is right, a generator object is returned</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">load_fasta</span><span class="p">(</span><span class="s">&quot;testdata.ffn&quot;</span><span class="p">)</span> 
<span class="go">&lt;generator object parse at ...&gt;</span>
</pre></div>
</div>
<p>The file must exist</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">load_fasta</span><span class="p">(</span><span class="s">&quot;yikes&quot;</span><span class="p">)</span>
<span class="go">Open: yikes  failed</span>
<span class="go">[Errno 2] No such file or directory: &#39;yikes&#39;</span>
</pre></div>
</div>
<p>However, we use Biopython and the fasta file&#8217;s syntax is not checked here!</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">load_fasta</span><span class="p">(</span><span class="s">&quot;testdata_fail.ffn&quot;</span><span class="p">)</span> 
<span class="go">&lt;generator object parse at ...&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="cobilib.load_fasta_from_url">
<tt class="descclassname">cobilib.</tt><tt class="descname">load_fasta_from_url</tt><big>(</big><em>url</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#load_fasta_from_url"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.load_fasta_from_url" title="Permalink to this definition">¶</a></dt>
<dd><p>Load fasta from a url</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>url:str</strong> :</p>
<blockquote>
<div><p>url to the fasta file</p>
</div></blockquote>
<p><strong>Returns</strong> :</p>
<p><strong>&#8212;&#8212;</strong> :</p>
<p><strong>fasta:Bio.SeqIO</strong> :</p>
<blockquote>
<div><p>fasta file parser object from Biopython
None, if error has occurred</p>
</div></blockquote>
<p><strong>Examples</strong> :</p>
<p class="last"><strong>&#8212;&#8212;-</strong> :</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="cobilib.load_genbank">
<tt class="descclassname">cobilib.</tt><tt class="descname">load_genbank</tt><big>(</big><em>filename</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#load_genbank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.load_genbank" title="Permalink to this definition">¶</a></dt>
<dd><p>loads a genbank file with filename as arg. RReturns generator object with list of gene</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>filename</strong> : str</p>
<blockquote>
<div><p>Path to the file to load</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>genes</strong> : Seq</p>
<blockquote class="last">
<div><p>Parser from Biopython</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="cobilib.load_plaintext">
<tt class="descclassname">cobilib.</tt><tt class="descname">load_plaintext</tt><big>(</big><em>filename</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#load_plaintext"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.load_plaintext" title="Permalink to this definition">¶</a></dt>
<dd><p>loads a nucleotide file with plaintext sequence. filename as arg. Returns list of genes</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.make_codon_histogram">
<tt class="descclassname">cobilib.</tt><tt class="descname">make_codon_histogram</tt><big>(</big><em>gene</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#make_codon_histogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.make_codon_histogram" title="Permalink to this definition">¶</a></dt>
<dd><p>returns codon_hist,aa_hist for one gene</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters :</th><td class="field-body"><p class="first"><strong>gene</strong> : SeqIO</p>
<blockquote>
<div><p>after loading fasta</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>codon_hist</strong> : dict with each gene.id as key and num_codonsx1 np.array for each codon</p>
<p class="last"><strong>aa_hist</strong> : dict with each gene.id as key and num_aax1 np.array for each aacid</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>The histogram should reproduce what www.kazusa.org.jp/codon/cgi-bin/countcodon.cgi computes
for the gene in testdata.ffn:</p>
<div class="highlight-python"><pre>UUU 12.6(     9)  UCU  9.8(     7)  UAU 15.4(    11)  UGU  4.2(     3)
UUC 21.1(    15)  UCC 16.9(    12)  UAC 11.2(     8)  UGC 12.6(     9)
UUA 15.4(    11)  UCA  8.4(     6)  UAA  0.0(     0)  UGA  1.4(     1)
UUG 12.6(     9)  UCG 12.6(     9)  UAG  0.0(     0)  UGG  5.6(     4)
CUU  8.4(     6)  CCU  4.2(     3)  CAU  9.8(     7)  CGU 22.5(    16)
CUC 15.4(    11)  CCC  5.6(     4)  CAC  7.0(     5)  CGC 25.3(    18)
CUA  4.2(     3)  CCA  1.4(     1)  CAA 11.2(     8)  CGA  2.8(     2)
CUG 53.4(    38)  CCG 23.9(    17)  CAG 19.7(    14)  CGG  5.6(     4)
AUU 36.5(    26)  ACU  5.6(     4)  AAU 28.1(    20)  AGU  7.0(     5)
AUC 16.9(    12)  ACC 23.9(    17)  AAC 19.7(    14)  AGC  8.4(     6)
AUA  0.0(     0)  ACA  0.0(     0)  AAA 26.7(    19)  AGA  0.0(     0)
AUG 29.5(    21)  ACG  9.8(     7)  AAG 14.0(    10)  AGG  1.4(     1)
GUU 25.3(    18)  GCU 25.3(    18)  GAU 37.9(    27)  GGU 28.1(    20)
GUC 19.7(    14)  GCC 32.3(    23)  GAC 18.3(    13)  GGC 28.1(    20)
GUA  8.4(     6)  GCA 12.6(     9)  GAA 47.8(    34)  GGA 12.6(     9)
GUG 32.3(    23)  GCG 35.1(    25)  GAG 18.3(    13)  GGG  9.8(     7)</pre>
</div>
<p>in the first field. Remember, the order of codons in this package is given in the variable codons</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">make_codon_histogram</span><span class="p">(</span> <span class="n">load_fasta</span><span class="p">(</span><span class="s">&quot;testdata.ffn&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">next</span><span class="p">()</span> <span class="p">)</span>
<span class="go">([0.012640449438202247, 0.021067415730337078, 0.015449438202247191, 0.012640449438202247, 0.0098314606741573031, 0.016853932584269662, 0.0084269662921348312, 0.012640449438202247, 0.015449438202247191, 0.011235955056179775, 0.0, 0.0, 0.0042134831460674156, 0.012640449438202247, 0.0014044943820224719, 0.0056179775280898875, 0.0084269662921348312, 0.015449438202247191, 0.0042134831460674156, 0.053370786516853931, 0.0042134831460674156, 0.0056179775280898875, 0.0014044943820224719, 0.023876404494382022, 0.0098314606741573031, 0.0070224719101123594, 0.011235955056179775, 0.019662921348314606, 0.02247191011235955, 0.025280898876404494, 0.0028089887640449437, 0.0056179775280898875, 0.036516853932584269, 0.016853932584269662, 0.0, 0.029494382022471909, 0.0056179775280898875, 0.023876404494382022, 0.0, 0.0098314606741573031, 0.028089887640449437, 0.019662921348314606, 0.026685393258426966, 0.014044943820224719, 0.0070224719101123594, 0.0084269662921348312, 0.0, 0.0014044943820224719, 0.025280898876404494, 0.019662921348314606, 0.0084269662921348312, 0.032303370786516857, 0.025280898876404494, 0.032303370786516857, 0.012640449438202247, 0.0351123595505618, 0.037921348314606744, 0.018258426966292134, 0.047752808988764044, 0.018258426966292134, 0.028089887640449437, 0.028089887640449437, 0.012640449438202247, 0.0098314606741573031], [0.033707865168539325, 0.10955056179775281, 0.063202247191011238, 0.026685393258426966, 0.0014044943820224719, 0.016853932584269662, 0.0056179775280898875, 0.0351123595505618, 0.016853932584269662, 0.030898876404494381, 0.05758426966292135, 0.053370786516853931, 0.029494382022471909, 0.039325842696629212, 0.047752808988764044, 0.040730337078651688, 0.085674157303370788, 0.10533707865168539, 0.056179775280898875, 0.066011235955056174, 0.078651685393258425])</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="cobilib.make_codon_histogram_dic">
<tt class="descclassname">cobilib.</tt><tt class="descname">make_codon_histogram_dic</tt><big>(</big><em>f</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#make_codon_histogram_dic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.make_codon_histogram_dic" title="Permalink to this definition">¶</a></dt>
<dd><p>returns codon_hist for all genes in a fasta file in form of a dictionary with fasta identifiers as keys</p>
<p>The histogram should reproduce what www.kazusa.org.jp/codon/cgi-bin/countcodon.cgi for the first gene in the testdata2.ffn:</p>
<div class="highlight-python"><pre>UUU 12.6(     9)  UCU  9.8(     7)  UAU 15.4(    11)  UGU  4.2(     3)
UUC 21.1(    15)  UCC 16.9(    12)  UAC 11.2(     8)  UGC 12.6(     9)
UUA 15.4(    11)  UCA  8.4(     6)  UAA  0.0(     0)  UGA  1.4(     1)
UUG 12.6(     9)  UCG 12.6(     9)  UAG  0.0(     0)  UGG  5.6(     4)
CUU  8.4(     6)  CCU  4.2(     3)  CAU  9.8(     7)  CGU 22.5(    16)
CUC 15.4(    11)  CCC  5.6(     4)  CAC  7.0(     5)  CGC 25.3(    18)
CUA  4.2(     3)  CCA  1.4(     1)  CAA 11.2(     8)  CGA  2.8(     2)
CUG 53.4(    38)  CCG 23.9(    17)  CAG 19.7(    14)  CGG  5.6(     4)
AUU 36.5(    26)  ACU  5.6(     4)  AAU 28.1(    20)  AGU  7.0(     5)
AUC 16.9(    12)  ACC 23.9(    17)  AAC 19.7(    14)  AGC  8.4(     6)
AUA  0.0(     0)  ACA  0.0(     0)  AAA 26.7(    19)  AGA  0.0(     0)
AUG 29.5(    21)  ACG  9.8(     7)  AAG 14.0(    10)  AGG  1.4(     1)
GUU 25.3(    18)  GCU 25.3(    18)  GAU 37.9(    27)  GGU 28.1(    20)
GUC 19.7(    14)  GCC 32.3(    23)  GAC 18.3(    13)  GGC 28.1(    20)
GUA  8.4(     6)  GCA 12.6(     9)  GAA 47.8(    34)  GGA 12.6(     9)
GUG 32.3(    23)  GCG 35.1(    25)  GAG 18.3(    13)  GGG  9.8(     7)</pre>
</div>
<p>in the first field. Remember, the order of codons in this package is given in the variable codons</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">make_codon_histogram_dic</span><span class="p">(</span> <span class="n">load_fasta</span><span class="p">(</span><span class="s">&quot;testdata2.ffn&quot;</span><span class="p">)</span> <span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">{&#39;fid|18348942|locus|VBIEscCol44059_0001|&#39;: [0.012640449438202247, 0.021067415730337078, 0.015449438202247191, 0.012640449438202247, 0.0098314606741573031, 0.016853932584269662, 0.0084269662921348312, 0.012640449438202247, 0.015449438202247191, 0.011235955056179775, 0.0, 0.0, 0.0042134831460674156, 0.012640449438202247, 0.0014044943820224719, 0.0056179775280898875, 0.0084269662921348312, 0.015449438202247191, 0.0042134831460674156, 0.053370786516853931, 0.0042134831460674156, 0.0056179775280898875, 0.0014044943820224719, 0.023876404494382022, 0.0098314606741573031, 0.0070224719101123594, 0.011235955056179775, 0.019662921348314606, 0.02247191011235955, 0.025280898876404494, 0.0028089887640449437, 0.0056179775280898875, 0.036516853932584269, 0.016853932584269662, 0.0, 0.029494382022471909, 0.0056179775280898875, 0.023876404494382022, 0.0, 0.0098314606741573031, 0.028089887640449437, 0.019662921348314606, 0.026685393258426966, 0.014044943820224719, 0.0070224719101123594, 0.0084269662921348312, 0.0, 0.0014044943820224719, 0.025280898876404494, 0.019662921348314606, 0.0084269662921348312, 0.032303370786516857, 0.025280898876404494, 0.032303370786516857, 0.012640449438202247, 0.0351123595505618, 0.037921348314606744, 0.018258426966292134, 0.047752808988764044, 0.018258426966292134, 0.028089887640449437, 0.028089887640449437, 0.012640449438202247, 0.0098314606741573031], &#39;fid|129049020348348942|locus|VBIEscCol44059_0001|&#39;: [0.012640449438202247, 0.021067415730337078, 0.015449438202247191, 0.012640449438202247, 0.0098314606741573031, 0.016853932584269662, 0.0084269662921348312, 0.012640449438202247, 0.015449438202247191, 0.011235955056179775, 0.0, 0.0, 0.0042134831460674156, 0.012640449438202247, 0.0014044943820224719, 0.0056179775280898875, 0.0084269662921348312, 0.015449438202247191, 0.0042134831460674156, 0.053370786516853931, 0.0042134831460674156, 0.0056179775280898875, 0.0014044943820224719, 0.023876404494382022, 0.0098314606741573031, 0.0070224719101123594, 0.011235955056179775, 0.019662921348314606, 0.02247191011235955, 0.025280898876404494, 0.0028089887640449437, 0.0056179775280898875, 0.036516853932584269, 0.016853932584269662, 0.0, 0.029494382022471909, 0.0056179775280898875, 0.023876404494382022, 0.0, 0.0098314606741573031, 0.028089887640449437, 0.019662921348314606, 0.026685393258426966, 0.014044943820224719, 0.0070224719101123594, 0.0084269662921348312, 0.0, 0.0014044943820224719, 0.025280898876404494, 0.019662921348314606, 0.0084269662921348312, 0.032303370786516857, 0.025280898876404494, 0.032303370786516857, 0.012640449438202247, 0.0351123595505618, 0.037921348314606744, 0.018258426966292134, 0.047752808988764044, 0.018258426966292134, 0.028089887640449437, 0.028089887640449437, 0.012640449438202247, 0.0098314606741573031]}</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="cobilib.make_codon_histogram_dic_combined">
<tt class="descclassname">cobilib.</tt><tt class="descname">make_codon_histogram_dic_combined</tt><big>(</big><em>f</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#make_codon_histogram_dic_combined"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.make_codon_histogram_dic_combined" title="Permalink to this definition">¶</a></dt>
<dd><p>make a codon and amino acid histogram from a file parser object. But this time combine all genes so that only
one histogram is returned.</p>
<dl class="docutils">
<dt>f:Bio.SeqIO</dt>
<dd>file parser object</dd>
</dl>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns :</th><td class="field-body"><p class="first"><strong>codon_hist:dict[&#8216;combined genome&#8217;] :: np.array x 64</strong> :</p>
<blockquote class="last">
<div><p>returns codon histogram</p>
</div></blockquote>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="cobilib.make_evolmatrix">
<tt class="descclassname">cobilib.</tt><tt class="descname">make_evolmatrix</tt><big>(</big><em>mutationmatrix</em>, <em>fitnessfunctions</em>, <em>fitnessmatrices</em>, <em>selection</em>, <em>additive=False</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#make_evolmatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.make_evolmatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Given the mutationmatrix, the fitnessfunctions, the fitnessmatrices (the amino-acid identity matrix) and the selection strength, this builds the evolutionmatrix</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.make_jc69_mutationmatrix">
<tt class="descclassname">cobilib.</tt><tt class="descname">make_jc69_mutationmatrix</tt><big>(</big><em>mu=0</em>, <em>alpha=0</em>, <em>beta=0</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#make_jc69_mutationmatrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.make_jc69_mutationmatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>ups, not correct! read models_of_dna_evolution on wiki</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.optimize_dic">
<tt class="descclassname">cobilib.</tt><tt class="descname">optimize_dic</tt><big>(</big><em>fasta_filename</em>, <em>fitfu</em>, <em>fitmat</em>, <em>minimal_number_of_nucleotides_per_gene=500</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#optimize_dic"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.optimize_dic" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>The probability when we  put r number of codons in a gene into n bins of different codons to find a codon with k occurences in the gene is
..math:</p>
<blockquote>
<div>p_k = left(</div></blockquote>
</div></blockquote>
<p>rac{r}{k})  
ight) 
rac{(n-1)^{r-k}}
rac{n^r}</p>
<blockquote>
<div><p>hence, the probability of finding more than one occurence is</p>
<dl class="docutils">
<dt>..math:</dt>
<dd>p_{k&gt;1} (63/64)^r( (64/63)^r - 1 )</dd>
</dl>
<p>if we want to to be reasonably sure our gene contains at least one codon from every kind we have to solve the inequality
p_{k&gt;1} &gt; p-value. A p-value of 0.95 implies we have to use a gene with at least 500 nucleotides</p>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="cobilib.optimize_sequence">
<tt class="descclassname">cobilib.</tt><tt class="descname">optimize_sequence</tt><big>(</big><em>steady_state</em>, <em>target_sequence</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#optimize_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.optimize_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>given the steady state and the target sequence, let us optimize!</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.relative_hellinger_distance">
<tt class="descclassname">cobilib.</tt><tt class="descname">relative_hellinger_distance</tt><big>(</big><em>steady_state</em>, <em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#relative_hellinger_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.relative_hellinger_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO: not implemented</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.relative_jeffrey_divergence">
<tt class="descclassname">cobilib.</tt><tt class="descname">relative_jeffrey_divergence</tt><big>(</big><em>steady_state</em>, <em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#relative_jeffrey_divergence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.relative_jeffrey_divergence" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO: not implemented</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.relative_minkowski_distance">
<tt class="descclassname">cobilib.</tt><tt class="descname">relative_minkowski_distance</tt><big>(</big><em>steady_state</em>, <em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#relative_minkowski_distance"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.relative_minkowski_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO: not implemented</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.remove_stopcodons_and_one_codon_amino_acids">
<tt class="descclassname">cobilib.</tt><tt class="descname">remove_stopcodons_and_one_codon_amino_acids</tt><big>(</big><em>codon_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#remove_stopcodons_and_one_codon_amino_acids"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.remove_stopcodons_and_one_codon_amino_acids" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="cobilib.run_model">
<tt class="descclassname">cobilib.</tt><tt class="descname">run_model</tt><big>(</big><em>parameters</em>, <em>fitnessfunctions</em>, <em>fitnessmatrices</em>, <em>aa_hist</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#run_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.run_model" title="Permalink to this definition">¶</a></dt>
<dd><p>parameters = alpha,beta,selection,t0,t1,t2,t3...</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.setup_parser">
<tt class="descclassname">cobilib.</tt><tt class="descname">setup_parser</tt><big>(</big><big>)</big><a class="reference internal" href="_modules/cobilib.html#setup_parser"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.setup_parser" title="Permalink to this definition">¶</a></dt>
<dd><p>Setting up the parser for command line usage, returns args</p>
</dd></dl>

<dl class="function">
<dt id="cobilib.spectral">
<tt class="descclassname">cobilib.</tt><tt class="descname">spectral</tt><big>(</big><em>hist</em>, <em>n_neighbors=10</em>, <em>n_components=2</em><big>)</big><a class="reference internal" href="_modules/cobilib.html#spectral"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#cobilib.spectral" title="Permalink to this definition">¶</a></dt>
<dd><p>docstring for spectral</p>
</dd></dl>

</div>
<span class="target" id="module-cobigui"></span></div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><em>Index</em></a></li>
<li><a class="reference internal" href="py-modindex.html"><em>Module Index</em></a></li>
<li><a class="reference internal" href="search.html"><em>Search Page</em></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to cobilib&#8217;s documentation!</a><ul>
<li><a class="reference internal" href="#brief">Brief</a></li>
<li><a class="reference internal" href="#summary">Summary</a></li>
<li><a class="reference internal" href="#routines">Routines</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/index.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="#">cobilib 1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Jan-Hendrik Trösemeier, Susanne Lipp, Christel Kamp.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>